### PR TITLE
TSForwardBase class and a transformation to reset `no_sinking` attribute

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
- Introduced a new TSForwardBase class for all TSForward transformations to simplify implementations and to reduce the amount of duplicate code.
- Added a new TSResetNoSinkingAttribute transformation to clean 'no_sinking' attribute after TSGeneral transformation execution.
- Fixed the issue when Transpose wasn't marked with 'no_sinking' attribute after TSForward transformation.
- Fixed the issue with legacy tensor names

Tests in progress

### Tickets:
 - *112680, 112386*

### CODE REVIEW

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 1]
```diff
@@ -23,22 +23,18 @@ using namespace ov::pass::transpose_sinking::utils;
...
-    ov::matcher_pass_callback matcher_pass_callback = [=](pattern::Matcher& m) {
-        const auto& pattern_to_output = m.get_pattern_map();
-
-        auto transpose = as_type_ptr<ov::op::v1::Transpose>(pattern_to_output.at(transpose_label));
-        auto main_node = as_type_ptr<ov::op::v8::Gather>(pattern_to_output.at(gather_label));
-        if (transformation_callback(main_node) || !main_node) {
+    auto sinking_transformation = [=](const std::shared_ptr<Node>& main_node,
+                                      const TransposeInputsInfo& transpose_info) -> bool {
+        auto gather = as_type_ptr<ov::op::v8::Gather>(main_node);
+        if (!main_node) {
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:31]: Did you mean this?
```suggestion
        auto gather = as_type_ptr<ov::op::v8::Gather>(main_node);
        if (!gather) {
```

File: src/common/transformations/src/transformations/transpose_sinking/ts_reset_no_sinking_attribute.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    MATCHER_SCOPE(TSResetNoSinkingAttribute);
+
+    auto transpose_label = wrap_type<ov::op::v1::Transpose>([](const Output<Node>& output) -> bool {
+        const auto& rt_info = output.get_rt_info();
+        return rt_info.find(NoTransposeSinkingAttr::get_type_info_static()) != rt_info.end();
+    });
+    ov::matcher_pass_callback matcher_pass_callback = [=](pattern::Matcher& m) {
+        const auto& pattern_to_output = m.get_pattern_map();
+
+        auto transpose = pattern_to_output.at(transpose_label);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_reset_no_sinking_attribute.cpp]: done

File: src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+protected:
+    static bool default_inputs_update(const std::shared_ptr<Node>& main_node,
+                                      const utils::TransposeInputsInfo& transpose_info);
+
+    void default_outputs_update(const std::shared_ptr<Node>& main_node,
+                                const utils::TransposeInputsInfo& transpose_info);
+
+private:
+    static bool IfNodeHasTransposeInputs(const Output<Node>& output,
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp (hunk 1 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp]: fixed

File: src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class ov::pass::transpose_sinking::TSForwardBase : public ov::pass::MatcherPass {
+public:
+    OPENVINO_RTTI("ov::pass::TSForwardBase", "0");
+    TSForwardBase() = default;
+
+    template <class... Types>
+    void create_pattern(bool const_transpose_input, const std::vector<size_t>& transpose_indices = {}) {
+        m_const_transpose_input = const_transpose_input;
+        m_tranpose_indices = transpose_indices;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp]: done

**Review by t-jankowski — APPROVED** [review]

**Review by mateusztabaka — APPROVED** [review]

File: src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                                          const TransposeInputsInfo& transpose_info) {
+    return utils::sink_forward::UpdateInputTransposes(main_node, transpose_info);
+}
+
+void TSForwardBase::default_outputs_update(const std::shared_ptr<Node>& main_node,
+                                           const TransposeInputsInfo& transpose_info) {
+    main_node->validate_and_infer_types();
+    for (auto& new_node : utils::sink_forward::InsertOutputTransposes(main_node, transpose_info)) {
+        register_new_node(new_node);
+        utils::UpdateForwardSinkingAbility(new_node);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp:67]: I believe we need to use it here

File: src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            res = sinking_transformation(main_node, transpose_input_info);
+        } else {
+            // default transpose sinking function:
+            res = default_inputs_update(main_node, transpose_input_info);
+            if (res) {
+                default_outputs_update(main_node, transpose_input_info);
+            }
+        }
+        if (!res) {
+            utils::UpdateForwardSinkingAbility(transpose_input_info.transpose);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp:48]: agree

File: src/common/transformations/src/transformations/transpose_sinking/ts_utils.cpp [hunk 7]
```diff
@@ -66,14 +66,22 @@ Output<Node> ChangeAxes(const Output<Node>& indices,
     return ChangeAxes(indices, data, axis);
 }
 
-TransposeInputsInfo GetFirstTransposeInput(const NodePtr& node) {
-    for (size_t input_idx = 0; input_idx < node->get_input_size(); ++input_idx) {
+TransposeInputsInfo GetFirstTransposeInput(const NodePtr& node,
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_utils.cpp:69]: I don't think so, we still need to check the next operations after new inserted Transposes

File: src/common/transformations/include/transformations/transpose_sinking/ts_reset_no_sinking_attribute.hpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+class TRANSFORMATIONS_API TSResetNoSinkingAttribute;
+
+}  // namespace transpose_sinking
+}  // namespace pass
+}  // namespace ov
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief TSGatherForward transformation sinks Transpose through Gather operations
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/transpose_sinking/ts_reset_no_sinking_attribute.hpp]: done

File: src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+class TRANSFORMATIONS_API TSForwardBase;
+
+}  // namespace transpose_sinking
+}  // namespace pass
+}  // namespace ov
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief TSForwardBase is a base class for all froward transformations.
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/transpose_sinking/ts_base.hpp]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_utils.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_utils.cpp:69]: Discussed offline, these functions can be deleted

File: src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_base.cpp]: fixed

**Review by evkotov — APPROVED** [review]

**Pull request merged**